sklearn/feature_selection/univariate_selection.py

```diff
@@ -227,6 +228,188 @@ def chi2(X, y):
     return _chisquare(observed, expected)
 
 
+def _info_gain(fc_count, c_count, f_count, fc_prob, c_prob, f_prob, total):
+    def get_c_f_cell(fc_prob, c_prob, f_prob):
+        t = np.log2(fc_prob/(c_prob * f_prob))
+        t[~np.isfinite(t)] = 0
+        return np.multiply(fc_prob, t)
+
+    def get_nc_nf_cell(fc_prob, c_prob, f_prob):
+        t = np.log2((1-f_prob-c_prob+fc_prob)/((1-c_prob)*(1-f_prob)))
+        t[~np.isfinite(t)] = 0
+        return np.multiply((1-f_prob-c_prob+fc_prob), t)
+
+    def get_c_nf_cell(c_prob, f_prob, c_count, fc_count, total):
+        nfc = (c_count - fc_count)/total
+        t = np.log2(nfc/(c_prob*(1-f_prob)))
+        t[~np.isfinite(t)] = 0
+        return np.multiply(nfc, t)
+
+    def get_nc_f_cell(c_prob, f_prob, f_count, fc_count, total):
+        fnc = (f_count - fc_count)/total
+        t = np.log2(fnc/((1-c_prob)*f_prob))
+        t[~np.isfinite(t)] = 0
+        return np.multiply(fnc, t)
+
+    # the feature score is averaged over classes
+    return (get_c_f_cell(fc_prob, c_prob, f_prob) +
+            get_nc_nf_cell(fc_prob, c_prob, f_prob) +
+            get_c_nf_cell(c_prob, f_prob, c_count, fc_count, total) +
+            get_nc_f_cell(c_prob, f_prob, f_count, fc_count, total)).mean(
+            axis=0)
+
+
+def get_feature_class_counts(X, y):
+    """Count feature, class, joint and total frequencies
+    """
+    X = check_array(X, accept_sparse=['csr', 'coo'])
+    if np.any((X.data if issparse(X) else X) < 0):
+        raise ValueError("Input X must be non-negative.")
+
+    Y = LabelBinarizer().fit_transform(y)
+    if Y.shape[1] == 1:
+        Y = np.append(1 - Y, Y, axis=1)
+
+    # counts
+
+    fc_count = safe_sparse_dot(Y.T, X)          # n_classes * n_features
+    total = fc_count.sum(axis=0).reshape(1, -1).sum()
+    f_count = X.sum(axis=0).reshape(1, -1)
+    if issparse(X):
+        c_count = (X.sum(axis=1).reshape(1, -1) * Y).T
```

Comment:
I also think this will return a matrix. To avoid any surprises, perhaps it should return an array.